### ISSUE TITLE
VS2015 & VS2017 Port Status

### ISSUE THREAD

**Issue by DarkStarSword** [bo3b/3Dmigoto#92]
VS2015:
- No known remaining issues
- Thoroughly tested for years
- Includes forward port to d3dcompiler_47
- Includes forward port to Windows 10 SDK (easily switched to any SDK from 8.1 and above, but cannot go back to SDK 8.0)
- Tested on Windows 7 without the platform update (note that using the newer SDK doesn't prevent this, but we do have to be careful not to introduce a link time dependency on newer d3d11/dxgi).
- Would be worthwhile using cmd_Decompiler to compare the assembler, disassembler and decompiler results in more games, as this has been one of the main areas of regressions due to minor changes in the disassembly output produced by d3dcompiler_47, as well as some libstdc++ formatting changes that the assembler was sensitive to. No issues remain in this area that I am aware of, but I would feel more confident after wider testing has been performed.

VS2017:
- crc32c implementation is completely broken - deal breaker until this is resolved:

    Creating cached resource [customshaderautoconvergence] resourceautoconvergencedepthdownscale128 = copy_desc resourcedepthbuffer
      NOTICE: cache now contains 20 resources (new hash: 0xd86d0c0b, sizeof(desc): 44)
      Resource Type = Texture2D
        Width = 128
        Height = 128
        MipLevels = 1
        ArraySize = 1
        Format = R32_FLOAT (41)
        SampleDesc.Count = 1
        SampleDesc.Quality = 0
        Usage = 0
        BindFlags = 0x28
        CPUAccessFlags = 0x0
        MiscFlags = 0x0
    80 00 00 00 80 00 00 00 01 00 00 00 01 00 00 00 29 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 28 00 00 00 00 00 00 00 00 00 00 00
    Creating cached resource [customshaderautoconvergence] resourceautoconvergencedepthdownscale128 = copy_desc resourcedepthbuffer
      NOTICE: cache now contains 21 resources (new hash: 0x67421c58, sizeof(desc): 44)
      Resource Type = Texture2D
        Width = 128
        Height = 128
        MipLevels = 1
        ArraySize = 1
        Format = R32_FLOAT (41)
        SampleDesc.Count = 1
        SampleDesc.Quality = 0
        Usage = 0
        BindFlags = 0x28
        CPUAccessFlags = 0x0
        MiscFlags = 0x0
    80 00 00 00 80 00 00 00 01 00 00 00 01 00 00 00 29 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 28 00 00 00 00 00 00 00 00 00 00 00

- Crash on exit in resource pool destructor while Releasing() cached resources. Possible refcounting bug - may well exist in master, but is exaccerbated by the above crc32c bug leading to the resource pool being overused.

**Comment by bo3b** [comment]
Weird.  Not the component I would predict to go south from the conversion.  I took a quick look at the crc32 code, and it's really pretty straightforward and simple.  I would not expect any C++ compiler difference to change this code.

Looking at the function going wrong here, this would be the spot I might expect to break with a change to a new C++ toolchain, which includes a new version of C++.  In that CommandList piece, the template is used as the source item, and it would not be too shocking for C++ to put a kink in this usage on an upgrade.  Maybe unfounded, but it is my impression that templates are a mess in C++.

**Comment by bo3b** [comment]
I ran a quick test with 2017 version, to step through the code to the crc32c function.  Just a small datapoint- it is calling the correct code, and using the CPU instruction version.  I wanted to be sure that __cpuid didn't change in some bizarre way, or that it went to the fallback table based path.

**Comment by DarkStarSword** [comment]
crc32 bug (actually, resource pool bug) resolved: https://github.com/bo3b/3Dmigoto/commit/43f9c201542bb8ba88468eb5c16a8c322da661f9

This was very much a matter of "how did this ever work?", but it turns out that the previous tool chains laid out the stack such that what we wanted to hash was in the area we were hashing, and the other bytes were always consistent, while vs2017 lays out the stack differently.

**Comment by DarkStarSword** [comment]
We potentially have a bit of an issue with ShaderRegex - any patterns matching the resource bind lines may be tripped up by the syntax changes there ("0" becomes "cb0", etc). The capitalisation of "cb" shouldn't be an issue - I intentionally made ShaderRegex case insensitive.

**Comment by DarkStarSword** [comment]
Now that we have a regression test suite for the HLSL decompiler a new issue has been uncovered in the vs2017 branch:

    FC4/9cc9ea03feed4639-vs.hlsl.chk
    -  r3.xyzw = (int4)r2.xyzw & int4(0,0,0,0);
    +  r3.xyzw = (int4)r2.xyzw & int4(65535,65535,65535,65535);
    
    JC3/f064c0539aaa5254-vs.hlsl.chk
    -  r2.xyz = (int3)r0.xzw & int3(0,0,255);
    +  r2.xyz = (int3)r0.xzw & int3(65535,65535,255);
    
    MirrorsEdge/00fe11cf0e780fea-cs.hlsl.chk
    -  r1.yzw = (int3)r0.xxx & int3(0,0,0);
    +  r1.yzw = (int3)r0.xxx & int3(65535,65535,65535);
    
    MirrorsEdge/53a20a1570726154-cs.hlsl.chk
    -  r1.yzw = (int3)r0.xxx & int3(0,0,0);
    +  r1.yzw = (int3)r0.xxx & int3(65535,65535,65535);
    
    MirrorsEdge/798abdcf8eeed9df-cs.hlsl.chk
    -  r1.yzw = (int3)r0.xxx & int3(0,0,0);
    +  r1.yzw = (int3)r0.xxx & int3(65535,65535,65535);
    ...
    -    r1.yz = (int2)r1.yz & int2(0,0);
    +    r1.yz = (int2)r1.yz & int2(65535,65535);
    
    MirrorsEdge/cdc7497832359d0d-cs.hlsl.chk
    -  r1.yzw = (int3)r0.xxx & int3(0,0,0);
    +  r1.yzw = (int3)r0.xxx & int3(65535,65535,65535);
    
    MirrorsEdge/e6db2e1b78527470-cs.hlsl.chk
    -  r1.yzw = (int3)r0.xxx & int3(0,0,0);
    +  r1.yzw = (int3)r0.xxx & int3(65535,65535,65535);
    
    Prey/24702f38f517e171-ps.hlsl.chk
    -    r1.xyz = (int3)r0.yzw & int3(0,0,0);
    -    r1.xyz = cmp((int3)r1.xyz != int3(0,0,0));
    +    r1.xyz = (int3)r0.yzw & int3(2139095040,2139095040,2139095040);
    +    r1.xyz = cmp((int3)r1.xyz != int3(2139095040,2139095040,2139095040));

**Comment by DarkStarSword** [comment]
This issue actually appears to be a bug in the *master* branch that is mysteriously fixed in the vs2017 branch.
Assembly:

    and r3.xyzw, r2.xyzw, l(0x0000ffff, 0x0000ffff, 0x0000ffff, 0x0000ffff)

Decompiled on master:

    r3.xyzw = (int4)r2.xyzw & int4(0,0,0,0); 

Decompiled on vs2017:

    r3.xyzw = (int4)r2.xyzw & int4(65535,65535,65535,65535);

**Comment by DarkStarSword** [comment]
Said bug is now fixed in master, preserving the hex literals, and this brings the vs2017 branch into parity with master as far as the HLSL regression tests are concerned.

**Comment by DarkStarSword** [comment]
New issue uncovered: The HLSL decompiler outputs "inf" on the vs2017 branch instead of "1.#INF". I personally prefer "inf", but we need to use what the compiler can parse. There are likely similar cases for NAN variants.

**Comment by usuarionuevor** [comment]
VS2019 when

**Comment by DarkStarSword** [comment]
Not for some time - there are a lot of build failures on vs2019 at present that will need to be resolved (typically these type of fixes can be incorporated into master provided they don't harm building on previous toolchain versions), but vs2019 will need extensive testing to ensure that any other weird bugs that crop up (and there will be) are squashed before we can even begin to consider switching master to it.

**Comment by DarkStarSword** [comment]
With Nintynuts looking to join the project I have finally flipped the switch and made vs2017 master. If there's any fallout we will deal with it as we would any other bug.

It may also worth looking at vs2019.
